lib/client.js

```diff
@@ -143,6 +156,58 @@ extend(Raven.prototype, {
     return this;
   },
 
+  makeErrorHandler: function () {
+    var self = this;
+    var caughtFirstError = false;
+    var caughtSecondError = false;
+    var calledFatalError = false;
+    var firstError;
+    return function (err) {
+      if (!caughtFirstError) {
+        // this is the first uncaught error and the ultimate reason for shutting down
+        // we want to do absolutely everything possible to ensure it gets captured
+        // also we want to make sure we don't go recursion crazy if more errors happen after this one
+        firstError = err;
+        caughtFirstError = true;
+        self.captureException(err, function (sendErr, eventId) {
+          if (!calledFatalError) {
+            calledFatalError = true;
+            self.onFatalError(err, sendErr, eventId);
+          }
+        });
+      } else if (calledFatalError) {
+        // we hit an error *after* calling onFatalError - pretty boned at this point, just shut it down
+        utils.consoleAlert('uncaught exception after calling fatal error shutdown callback - this is bad! forcing shutdown');
```

Comment:
I think we should talk about this more next week but it's fine for now. nice PR 👍